src/core/varlink-unit.c

```diff
@@ -0,0 +1,449 @@
+                        JSON_BUILD_PAIR_DUAL_TIMESTAMP_NON_NULL("AssertTimestamp", &u->assert_timestamp),
+                        SD_JSON_BUILD_PAIR_CONDITION(!sd_id128_is_null(u->invocation_id), "InvocationID", SD_JSON_BUILD_UUID(u->invocation_id)),
+                        JSON_BUILD_PAIR_CALLBACK_NON_NULL("Markers", markers_build_json, &u->markers),
+                        JSON_BUILD_PAIR_CALLBACK_NON_NULL("ActivationDetails", activation_details_build_json, u->activation_details),
+                        SD_JSON_BUILD_PAIR_BOOLEAN("DebugInvocation", u->debug_invocation));
+}
+
+static int list_unit_one(sd_varlink *link, Unit *unit, bool more) {
+        _cleanup_(sd_json_variant_unrefp) sd_json_variant *v = NULL;
+        int r;
+
+        assert(link);
+        assert(unit);
+
+        r = sd_json_buildo(
+                &v,
+                SD_JSON_BUILD_PAIR_CALLBACK("Context", unit_context_build_json, unit),
+                SD_JSON_BUILD_PAIR_CALLBACK("Runtime", unit_runtime_build_json, unit));
+        if (r < 0)
+                return r;
+
+        if (more)
+                return sd_varlink_notify(link, v);
+
+        return sd_varlink_reply(link, v);
+}
+
+static int lookup_unit_by_name(sd_varlink *link, Manager *manager, const char *name, Unit **ret_unit) {
+        assert(link);
+        assert(manager);
+        assert(ret_unit);
+
+        if (!name)
+                return -EINVAL;
+        if (!unit_name_is_valid(name, UNIT_NAME_PLAIN|UNIT_NAME_INSTANCE))
+                return -EINVAL;
+
+        Unit *u = manager_get_unit(manager, name);
+        if (!u)
+                return -ESRCH;
+
+        *ret_unit = u;
+        return 0;
+}
+
+static int lookup_unit_by_pidref(sd_varlink *link, Manager *manager, PidRef *pidref, Unit **ret_unit) {
+        _cleanup_(pidref_done) PidRef peer = PIDREF_NULL;
+        Unit *unit;
+        int r;
+
+        assert(link);
+        assert(manager);
+        assert(ret_unit);
+
+        if (pidref_is_automatic(pidref)) {
+                r = varlink_get_peer_pidref(link, &peer);
+                if (r < 0)
+                        return log_debug_errno(r, "Failed to get peer pidref: %m");
+
+                pidref = &peer;
+        } else if (!pidref_is_set(pidref))
+                return -EINVAL;
+
+        unit = manager_get_unit_by_pidref(manager, pidref);
+        if (!unit)
+                return -ESRCH;
+
+        *ret_unit = unit;
+        return 0;
+}
+
+typedef struct UnitLookupParameters {
+        const char *name;
+        PidRef pidref;
+} UnitLookupParameters;
+
+static void unit_lookup_parameters_done(UnitLookupParameters *p) {
+        assert(p);
+        pidref_done(&p->pidref);
+}
+
+int vl_method_list_units(sd_varlink *link, sd_json_variant *parameters, sd_varlink_method_flags_t flags, void *userdata) {
+        static const sd_json_dispatch_field dispatch_table[] = {
+                // XXX name or id???
```

Comment:
@DaanDeMeyer, there is also this question.

Reply:
I'd do the same in varlink, name as input

Reply:
Please drop this comment